clients/src/main/java/org/apache/kafka/clients/consumer/internals/DefaultBackgroundThreadRunnable.java

```diff
@@ -0,0 +1,284 @@
+        this.metadata = metadata;
+        this.networkClient = client;
+        this.metrics = new Metrics();
+        this.running = true;
+    }
+
+    private void setConfig() {
+        this.retryBackoffMs = this.config.getLong(ConsumerConfig.RETRY_BACKOFF_MS_CONFIG);
+        this.clientId = config.getString(CommonClientConfigs.CLIENT_ID_CONFIG);
+        this.heartbeatIntervalMs = config.getInt(ConsumerConfig.HEARTBEAT_INTERVAL_MS_CONFIG);
+    }
+
+    @Override
+    public void run() {
+        try {
+            log.debug("{} started", getClass());
+            while (running) {
+                runOnce();
+                time.sleep(retryBackoffMs);
+            }
+        } catch (InterruptException e) {
+            log.error("The background thread has been interrupted");
+            this.exception.set(Optional.of(new RuntimeException(e)));
+        } catch (Throwable t) {
+            log.error("The background failed due to unexpected error", t);
+            if (t instanceof RuntimeException)
+                this.exception.set(Optional.of((RuntimeException) t));
+            else
+                this.exception.set(Optional.of(new RuntimeException(t)));
+        } finally {
+            close();
+            log.debug("{} closed", getClass());
+        }
+    }
+
+    /**
+     * Process event from a single poll
+     */
+    void runOnce() {
+        this.inflightEvent = maybePollEvent();
+        log.debug("processing applicatoin event: {}", this.inflightEvent);
+        if (this.inflightEvent.isPresent() && maybeConsumeInflightEvent(this.inflightEvent.get())) {
+            // clear inflight event upon successful consumption
+            this.inflightEvent = Optional.empty();
+        }
+        networkClient.pollNoWakeup();
+    }
+
+    public Optional<ApplicationEvent> maybePollEvent() {
+        if (this.inflightEvent.isPresent() || this.applicationEventQueue.isEmpty()) {
+            return this.inflightEvent;
+        }
+        return Optional.ofNullable(this.applicationEventQueue.poll());
+    }
+
+    /**
+     * ApplicationEvent are consumed here.
+     * @param event an {@link ApplicationEvent}
+     * @return true when successfully consumed the event.
+     */
+    public boolean maybeConsumeInflightEvent(ApplicationEvent event) {
+        log.debug("try consuming event: {}", Optional.ofNullable(event));
+        switch (event.type) {
+            case NOOP:
+                process((NoopApplicationEvent) event);
+                return true;
+            default:
+                inflightEvent = Optional.empty();
+                log.warn("unsupported event type: {}", event.type);
+                return true;
+        }
+    }
+
+    /**
+     * Processes {@link NoopApplicationEvent} and equeue a {@link NoopBackgroundEvent}. This is intentionally left here
+     * for demonstration purpose.
+     * @param event a {@link NoopApplicationEvent}
+     */
+    private void process(NoopApplicationEvent event) {
```

Comment:
This seems like what a mock client needs to do, why we need to add it to the default implementation? Or are we going to remove it in this class later?